File: getting-started/variant-go/module-05-functions-and-decomposition/exercise-03-file-organization/solution/validate.go
package main

import (
	"fmt"
	"strings"
)

func validateEmail(email string) bool {
	return strings.Contains(email, "@") && strings.Contains(email, ".")
}

func validateStudent(s Student) []string {
	var errors []string
	if s.Name == "" {
		errors = append(errors, "name is required")
	}
	if !validateEmail(s.Email) {
		errors = append(errors, "invalid email")
	}
	if len(s.Scores) == 0 {
		errors = append(errors, "at least one score required")
	}
	for _, score := range s.Scores {
		if score < 0 || score > 100 {
			errors = append(errors, fmt.Sprintf("score %d out of range", score))
		}
	}
	if s.Tuition < 0 {
		errors = append(errors, "tuition cannot be negative")
	}
	return errors
}

File: getting-started/variant-ts/module-05-functions-and-decomposition/exercise-03-file-organization/solution/validate.js
const validateEmail = (email) => {
  return email.includes("@") && email.includes(".");
};

export const validateStudent = (student) => {
  const errors = [];
  if (student.name === "") {
    errors.push("name is required");
  }
  if (!validateEmail(student.email)) {
    errors.push("invalid email");
  }
  if (student.scores.length === 0) {
    errors.push("at least one score required");
  }
  for (const score of student.scores) {
    if (score < 0 || score > 100) {
      errors.push(`score ${score} out of range`);
    }
  }
  if (student.tuition < 0) {
    errors.push("tuition cannot be negative");
  }
  return errors;
};
